javaslang/src/main/java/javaslang/collection/ArrayType.java

```diff
@@ -0,0 +1,147 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import java.io.Serializable;
+
+abstract class ArrayType<T> implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    @SuppressWarnings("unchecked")
+    static <T> ArrayType<T> obj() { return (ArrayType<T>) ObjectArrayType.INSTANCE; }
+
+    abstract Class<T> type();
+    abstract int lengthOf(Object array);
+    abstract T getAt(Object array, int index);
+
+    abstract Object empty();
+    abstract void setAt(Object array, int index, Object value);
+    abstract Object copy(Object array, int arraySize, int sourceFrom, int destinationFrom, int size);
+
+    @SuppressWarnings("unchecked")
+    static <T> ArrayType<T> of(Object array) { return of((Class<T>) array.getClass().getComponentType()); }
+    @SuppressWarnings("unchecked")
+    private static <T> ArrayType<T> of(Class<T> type) {
+        if (!type.isPrimitive()) {
+            return (ArrayType<T>) obj();
+        } else {
+            if (boolean.class == type) {
+                return (ArrayType<T>) BooleanArrayType.INSTANCE;
+            } else if (byte.class == type) {
+                return (ArrayType<T>) ByteArrayType.INSTANCE;
+            } else if (char.class == type) {
+                return (ArrayType<T>) CharArrayType.INSTANCE;
+            } else if (double.class == type) {
+                return (ArrayType<T>) DoubleArrayType.INSTANCE;
+            } else if (float.class == type) {
+                return (ArrayType<T>) FloatArrayType.INSTANCE;
+            } else if (int.class == type) {
+                return (ArrayType<T>) IntArrayType.INSTANCE;
+            } else if (long.class == type) {
+                return (ArrayType<T>) LongArrayType.INSTANCE;
+            } else if (short.class == type) {
+                return (ArrayType<T>) ShortArrayType.INSTANCE;
+            } else {
+                throw new IllegalArgumentException("Unknown type: " + type);
+            }
+        }
```

Comment:
I wanted to separate the primitive types from the only non-primitive type.
Would you like me to merge the nested ifs instead?